ucp/_libs/ucx_api.pyx

```diff
@@ -0,0 +1,475 @@
+            pass
+
+    @property
+    def handle(self):
+        return int(<uintptr_t><void*>self._handle)
+
+    def request_cancel(self, ucp_request_as_int):
+        cdef ucp_request *req = <ucp_request*><uintptr_t>ucp_request_as_int
+
+        # Notice, `ucp_request_cancel()` calls the send/recv callback function,
+        # which will handle the request cleanup.
+        ucp_request_cancel(self._handle, req)
+
+    def tag_recv(self, buffer, size_t nbytes, ucp_tag_t tag, cb_func, cb_args):
+        cdef void *data = <void*><uintptr_t>(get_buffer_data(
+            buffer, check_writable=True)
+        )
+        cdef ucp_tag_recv_callback_t _tag_recv_cb = (
+            <ucp_tag_recv_callback_t>_ucx_tag_recv_callback
+        )
+        cdef ucs_status_ptr_t status = ucp_tag_recv_nb(self._handle,
+                                                       data,
+                                                       nbytes,
+                                                       ucp_dt_make_contig(1),
+                                                       tag,
+                                                       -1,
+                                                       _tag_recv_cb)
+        return handle_comm_result(
+            status, {"cb_func": cb_func, "cb_args": cb_args}, expected_receive=nbytes
+        )
+
+
+cdef class UCXEndpoint:
+    """Python representation of `ucp_ep_h`
+    Please use `ucx_ep_create()` to contruct an instance of this class
+    """
+    cdef:
+        ucp_ep_h _handle
+        bint initialized
+
+    def close(self, UCXWorker worker):
+        cdef ucs_status_ptr_t status
+        if self.initialized:
+            status = ucp_ep_close_nb(self._handle, UCP_EP_CLOSE_MODE_FLUSH)
+            self.initialized = False
+            if UCS_PTR_STATUS(status) != UCS_OK:
+                assert not UCS_PTR_IS_ERR(status)
+                # We spinlock here until `status` has finished
+                while ucp_request_check_status(status) != UCS_INPROGRESS:
+                    worker._worker.progress()
+                assert not UCS_PTR_IS_ERR(status)
+                ucp_request_free(status)
+
+    def info(self):
+        assert self.initialized
+        # Making `ucp_ep_print_info()` write into a memstream,
+        # convert it to a Python string, clean up, and return string.
+        cdef char *text
+        cdef size_t text_len
+        cdef FILE *text_fd = open_memstream(&text, &text_len)
+        assert(text_fd != NULL)
+        ucp_ep_print_info(self._handle, text_fd)
+        fflush(text_fd)
+        cdef unicode py_text = text.decode()
+        fclose(text_fd)
+        free(text)
+        return py_text
+
+    @property
+    def handle(self):
+        assert self.initialized
+        return int(<uintptr_t><void*>self._handle)
+
+
+cdef UCXEndpoint ucx_ep_create(ucp_ep_h ep):
+    ret = UCXEndpoint()
+    ret._handle = ep
+    ret.initialized = True
+    return ret
```

Comment:
Since you're making this the "Pythonic" side of things, wouldn't it make sense that creating a `UCXEndpoint` from a class instance suffice and not requiring `ucx_ep_create` anymore?

Reply:
Did this get resolved @madsbk?